core/blockchain.go

```diff
@@ -1899,8 +1900,7 @@ func (bc *BlockChain) reprocessState(current *types.Block, reexec uint64) error
 		// Flatten snapshot if initialized, holding a reference to the state root until the next block
 		// is processed.
 		if err := bc.flattenSnapshot(func() error {
-			triedb.Reference(root, common.Hash{})
-			if previousRoot != (common.Hash{}) {
+			if previousRoot != (common.Hash{}) && previousRoot != root {
```

Comment:
must check that previousRoot != root to avoid an edge case in reprocessing blocks when the parent is the same as the child, as `triedb.Reference` is not called the 2nd time (during block processing through statedb.Commit, because of a check there with root != origin)